### ISSUE TITLE
Support freeze/thaw

### ISSUE THREAD

**Issue by behlendorf** [openzfs/zfs#260]
ZFS has hooks for suspending the filesystem but they have not yet been integrated with their Linux freeze/thaw counterparts.  This must be done before you can safely hibernate a system running ZFS.

**Comment by ccic** [comment]
@kohlschuetter and @behlendorf , the above implementation has two issues:
1. When the call of zfs_suspend_fs returns, the caller still holds two locks: 'z_teardown_lock' and 'z_teardown_inactive_lock'. The process cannot exit withholding those locks, otherwise it caused issues. So, a possible modification is call zfs_suspend_fs and sleep for a specified duration, and then calling zfs_resume_fs to release those locks. In other words, calling zfs_suspend_fs and zfs_resume_fs in pairs in order to apply and release locks in a correct way.
2. zfs_suspend_fs/zfs_resume_fs is not sufficient to freeze/thaw the pool. Consider a scenario, some file systems share one pool, if you suspend the writes from some of the file system, the other file system are still allowed to write to the pool. Moreover, even the file system is suspended, the property setting of 
@kohlschuetter and @behlendorf , the above implementation has two issues:
1. When the call of zfs_suspend_fs returns, the caller still holds two locks: 'z_teardown_lock' and 'z_teardown_inactive_lock'. The process cannot exit withholding those locks, otherwise it caused issues. So, a possible modification is call zfs_suspend_fs and sleep for a specified duration, and then calling zfs_resume_fs to release those locks. In other words, calling zfs_suspend_fs and zfs_resume_fs in pairs in order to apply and release locks in a correct way.
2. zfs_suspend_fs/zfs_resume_fs is not sufficient to freeze/thaw the pool. Consider a scenario, some file systems share one pool, if you suspend the writes from some of the file system, the other file system are still allowed to write to the pool. Moreover, even the file system is suspended, the property setting of pool is still allowed. So, that is not a "real" freezing of the disk. We should explore other methods, for example, freezing uberlock, to meet the "real" freezing feature.

**Comment by behlendorf** [comment]
@ccic thanks for taking the time to investigate how this functionality could be implemented.  The changes proposed here should be considered an initial prototype WIP to help us investigate the various issues.  Unfortunately, adding this functionality hasn't been a hasn't priority for the team.  Regarding you specific concerns.

> 1. When the call of zfs_suspend_fs returns, the caller still holds two locks: 'z_teardown_lock' and 'z_teardown_inactive_lock'. The process cannot exit withholding those locks,

Good point. So one possible avenue worth exploring might be to have a freeze take a snapshot and then use the existing rollback code to effectively pivot on to that snapshot.  That would allow us to use the existing model of suspend/rollback/resume except that you'd be resuming on an immutable snapshot.

> 2. zfs_suspend_fs/zfs_resume_fs is not sufficient to freeze/thaw the pool. 

Using a snapshot would provide a solid guarantee on immutability.  As for allowing the pool to be manipulated or other non-frozen filesystems it's not clear that's a problem.  The VFS is only requesting that a specific super block be frozen.  If freezing an entire pool is needed than alternate interfaces will be needed.

**Comment by ccic** [comment]
@behlendorf thanks for your sharing of your consideration. I know this feature is not a priority. I just want to get some clues about how to design and implement it. 
For the existing model of suspend/rollback/resume, I have checked the code, the zfs_ioc_rollback has already contained the logic of suspend and then resume, so a possible avenue is to (1) take a snapshot of a specified file system, (2) rollback it but wait for a while after suspending (here we freeze the fs), then (3) resume it Yes it has the effect of freezing file system. Am I correct?

**Comment by behlendorf** [comment]
@ccic yes, it should have that effect.

**Comment by isegal** [comment]
+1
Definitely, would love to have this feature for backup snapshot scenarios in the cloud.

For example on AWS one can perform a point-in-time snapshot of an attached EBS drive. Some backup tools rely on FS flushing and freezing so that the snapshot data is consistent. For example with xfs_freeze we are able to snapshot a raid array with no consistency issues.

An example of this is the mysql-ebs-backup script that's currently tailored for XFS on EBS: https://github.com/graphiq/mysql-ebs-snapshot.

If anyone knows for a workaround (i.e. sync command perhaps)? Please do share.

**Comment by ccic** [comment]
A quick workaround for one ZFS FS may be:
(1) Expose zfs_suspend_fs/zfs_resume_fs to users through zfs command. That only takes less than 100 lines, I think. Since zfs_suspend_fs still holds two locks, so it requires users to specify another input: how long do you want to suspend? for example, 5 seconds (I think it is long enough)
(2) Execute "zfs suspend 5" ('suspend' is the command, 5 means 5 seconds) to freeze
(3) The other logic on MySQL ..
(4) Resume the ZFS after suspend timeout automatically.

For a complete feature to freeze zpool, we have to (1) flush dirty pages and (2) suspend write to the pool. But now, it prevents from being written to disk, unfortunately it excludes synchronous write. It still needs more investigation.

**Comment by isegal** [comment]
Does /bin/sync flush dirty ZFS pages to disk?

**Comment by ccic** [comment]
'sync' asks file system to flush the pending write to disk, but I'm not sure about the answer.

**Comment by MyPod** [comment]
@isegal no, sync doesn't guarantee that data lands on disk, only that it reaches at least the zil. The issue on O_DIRECT should explain why, iirc.

**Comment by gasparch** [comment]
Hi all! 

So what is a progress on this ? I'm trying to setup tuxonice + ubuntu root/home over zfs + LUKS to encrypt everything under zfs.

If I use zfs as a root fs for system and execute hibernate to some other swap partition (which saves all kernel memory + some page cache) what potentially can go wrong with internal zfs structures? Even if there is data to be written to main disk - it will be in zil and will be saved as part of memory during hibernation process, isn't it?

If system fails to restore from hibernated state - that would be equal to poweroff from OS point of view - which is totally possible scenario during normal work as well and I'm not sure I want to protect from this one.

If hibernation process restores properly and continues execution - how does it hurt ZFS? What structures can be corrupt? Do ZFS rely on kernel page cache (because this one is partially cleaned during hibernation)?

**Comment by candlerb** [comment]
The primary application for me would be taking live block-level backups of virtual machine images, where the VM (guest) is using ZFS.

*In theory* if I took a snapshot while it was in the middle of flushing a transaction group or writing the ZIL to disk, the filesystem image would still be consistent and recoverable.  I would be happier if I didn't have to trust to this.

There is a Qemu hook to "quiesce" a guest's filesystem before snapshotting it, by means of an agent running in the guest: see

* https://wiki.libvirt.org/page/Qemu_guest_agent
* https://wiki.qemu.org/Features/GuestAgent
* https://specs.openstack.org/openstack/nova-specs/specs/kilo/implemented/quiesced-image-snapshots-with-qemu-guest-agent.html
* https://github.com/sheepdog/sheepdog/wiki/How-to-make-a-live-snapshot-properly

It appears that under the good, `guest-fsfreeze-freeze`just calls fsfreeze, presumably on all mounted filesystems.  If a guest has ext4 for root and zfs for `/data` I'd like the call to succeed, i.e. saying that it successfully froze all filesystems.

**Comment by hermes-pimentel** [comment]
behlendorf opened this issue on May 31, 2011...

**Comment by candlerb** [comment]
@gasparch:

> Even if there is data to be written to main disk - it will be in zil and will be saved as part of memory during hibernation process, isn't it?

No: the ZIL is only ever used for synchronous writes.

What happens in normal operation is that ZFS batches up writes in RAM into a [transaction group](https://www.delphix.com/blog/delphix-engineering/zfs-fundamentals-transaction-groups), and then flushes that transaction group to disk.  Since ZFS doesn't overwrite any blocks in-place, you can in principle pull the plug at any point, and see either the filesystem state entirely as it was prior to starting to write the TG, or after the complete successful write of the TG.  So at least the filesystem state is always consistent. (*)

But what happens to the blocks in an open transaction group during a hibernate/suspend, I don't know.

(*) This should be safe as long as the underlying disk isn't doing a scary amount of write reordering, such that the [uberblock is written out](https://utcc.utoronto.ca/~cks/space/blog/solaris/ZFSHowMetadataUpdates) *before* the blocks it points to.

**Comment by dankimmel** [comment]
Actually, you'd also want step 5: after restoring and doing `import --restore-from-checkpoint`, you'd also want to discard the checkpoint with `zpool checkpoint --discard` (for the same reason as in step 3)

**Comment by eblau** [comment]
Wow, I did not realize this support was missing. I run ZFS as a root filesystem on a laptop. I've been successfully hibernating and resuming for years up until I upgraded to Linux 5.7.7 with ZFS 0.8.4. I've corrupted my zpool twice in the past month from hibernating and resuming with these versions.

I presume this is "expected" since these hooks are not present for ZFS. We should never hibernate and resume with ZFS?

**Comment by xuanruiqi** [comment]
Hmm, so is this issue the only thing blocking hibernation support for ZoL?

**Comment by cytrinox** [comment]
Is the missing implementation only problematic for root on zfs or for every scenario? I want to build a 8x4TB raidz2 on my new workstation where root is on a mdraid/ext4 but I need hibernate (suspend2disk) every day.

**Comment by AttilaFueloep** [comment]
Just a random data point: I'm using hibernation with root on zfs for years without any issue. Thanks to arch Linux and zfs-git always on current versions. Same use case: notebook, at least daily hibernated. It wouldn't hurt to checkpoint the pool before hibernating though.

**Comment by 8573** [comment]
As a different data point, in 2015 or 2016 and possibly as recently as 2019 (looking at Git history), I occasionally accidentally would suspend my laptop with root on ZFS, because logind would default to doing so on certain hardware events (esp. lid close), and this would result in ZFS-related error messages printed on the kernel console that would frighten me into trying harder to configure logind not to do that and to remember not to close the lid all the way.

**Comment by misterbigstuff** [comment]
^ hibernate and suspend are NOT the same thing. suspend is safe, hibernate is not - and doesn't work, anyway, because required initrd code isn't there.

**Comment by Greek64** [comment]
@misterbigstuff 
I most definitely use **hibernate** on a daily basis (with the occasional clean reboot every 1-2 weeks) on a zfs-on-root debian testing system for about 1.5 years now without any problems...

**Comment by misterbigstuff** [comment]
with swap on ZFS? no non-ZFS partitions?

**Comment by Greek64** [comment]
Well yes, the swap is on a separate non-zfs LUKS encrypted partition.
But this issue is about freeze/thaw support for hibernation (Not hibernating using swap on zfs)

**Comment by misterbigstuff** [comment]
there are multiple issues. pure ZFS hibernate will not work because of both issues, but your setup will occasionally fail because of just this one.

**Comment by eblau** [comment]
@Greek64 @AttilaFueloep I was in the same situation as you. Daily hibernates with a LUKS encrypted ZFS root and a non-ZFS LUKS swap partition was working fine for me for years. Then I got multiple zpool corruptions within a month of each other and came across this GitHub issue. I would strongly recommend not hibernating. Suspend works perfectly. If you do continue using hibernate, be sure to have a recent backup.

**Comment by AttilaFueloep** [comment]
@eblau I really appreciate your recommendation and I know that I'm operating outside specifications here. What I do to protect against corruption is to run `zpool checkpoint` in the process of hibernating. The hope is to be able to rewind to this checkpoint, should to the pool go corrupt. I've recent file and snapshot based backups as well of course, not least because I'm running the zfs master branch. So I think I'm going to continue to use hibernation unless bitten by pool corruption (once bitten twice shy ;-).

By the way, our setups differ in mine having root on zfs and just `/boot` and swap on LUKS. maybe this makes a difference?

**Comment by kloczek** [comment]
Someone can summarise current state of this issue?
What is dine?
What needs to be done?
Something else?

**Comment by asdf8dfafjk** [comment]
Since no reply, I can tell you something that definitely fails: zfs on root, hibernate it to disk. Upon turning on, if the same root is booted again (which will eventually start wake up of the hibernated state) then there is a chance your system will go kaput. (I've had 4 zfs dataset losses this way).

Extremely simple solution:

I now have two installations (both share swap partition)- 1 where I do main work. Which I keep suspending all the time. Second which I use for main grub entry and which is the one that actually wakes up my original work (thus avoiding an import of the pool)

**Comment by luke-jr** [comment]
Ugh, shouldn't ZFS at least inhibit hibernation so long as this is a problem? (So the hibernation attempts fail instead of corrupting stuff)

**Comment by xuanruiqi** [comment]
I agree. If root is on ZFS, ZFS should probably inhibit hibernation.

**Comment by asdf8dfafjk** [comment]
@luke-jr @xuanruiqi 

Guys I use hibernate too and sorry if it's rude but I dont think it's appropriate to use "should" for volunteer work. These guys are giving away such an amazing (in the true sense of the word, schmucks like me now have mirrors so I dont care about hdd crash) piece of software for free and telling them what should or should not happen doesn't seem appropriate. 

Perhaps you could start a new issue and discuss design for your request and consider a PR?

**Comment by Atemu** [comment]
Discussing what *should* be done about an issue, including not immediately obvious short-term mitigations, is in no way disrespectful.  
OpenZFS is a FOSS project and therefore given to us "for free" but that does not mean that it's free of issues. Discussing these is absolutely fine (and necessary IMO) as long as it's done constructively and in a nice tone.

**Comment by asdf8dfafjk** [comment]
Sorry but your point makes no sense. I do not claim that there are no issues. But posting repeated iterations of "this is a must", "I need this", "what is the status on this", " <so and so> should be done" does not help progress the issue specially coming from people who have made 0 contributions to this project. (In fact when I saw the first "must" comment my first thought was someone speaking so authoritatively must be an active member, but alas, he has 0 pulls).

IMHO once the issue is known, the only valid contribution is identifying and discussing a fix, as some other kind members have already done and @behlendorf seems to be very prompt in responses 

Let's not bother the maintainers anymore with noise. I have already proposed a workaround. If you're too lazy to use it, please dont expect maintainers to do the work for you.

**Comment by dm17** [comment]
Trying to debug if this laptop's inability to wake from s2ram (using zzz utility) is due to its use of ZFS (on rootfs) - if anyone has ideas :)
Thanks!

**Comment by Atemu** [comment]
Hibernation yes, suspension no.

**Comment by danielmorlock** [comment]
Using hibernation results in corrupted zpools sooner or later. That tells me that there might be an issue with freezing or thawing: I just used a corrupted zpool for weeks without knowing, since hibernation always resumed a proper zpool state so that even "zpool status" reports no errors in the resumed system. After a reboot, I was not able to import the zpool because of:

```
[root@sysrescue /mnt]# zpool import -f -R /mnt/rpool_ws1 -o readonly=on  rpool_ws1
cannot import 'rpool_ws1': I/O error
	Destroy and re-create the pool from
	a backup source.
```

So there should be either a more prominent warning in the docs that warns user from using ZFS on a desktop pc or the hibernation process should be cancelled from ZFS. The current state is dangerous since users can fall into this trap and can loose data.

**Comment by bghira** [comment]
ZFS has no control over this.

**Comment by bghira** [comment]
also, even if the distribution disables hibernate on ZFS, people just go out of their way to avoid all warnings and do whatever they want: https://askubuntu.com/questions/1266599/how-to-increase-swap-with-zfs-on-ubuntu-20-04-to-enable-hibernation

**Comment by eblau** [comment]
@bghira It doesn't seem believable to me that ZFS has no control over this. Surely there are callbacks that can be registered with the kernel to be invoked when hibernate is invoked. At the very least these could be implemented to hang or kernel panic instead of allowing hibernate to proceed and silently corrupt the zpool.

**Comment by bghira** [comment]
as i understand, they're GPL-only symbols.

**Comment by danielmorlock** [comment]
After digging into the kernel hibernation (suspend to disk) process with @problame, we figured out that ZFS (event with hibernate) did not cause the  zpool corruption. Further during our (rough) analyze, we did not found a reason why ZFS wouldn't work with hibernation iff the swap is outside of ZFS. Even without the `freeze_fs` hooks mentioned by @kohlschuetter, hibernating ZFS should "just work". I guess the hooks are relevant for doing hibernation to a swap that is inside the ZFS.

TL;DR: The problem was in `genkernel`(Gentoo automatic kernel building scripts) that includes a script for initramfs. This script is doing luks-encryption and boots from what is listed in the boot options. In the case of ZFS, I have a crypted swap and a crypted root including the ZFS pool. The initramfs script decrypts the root and imports the zpool BEFORE if decrypts the SWAP including the RAM state for hibernation. So the pool is always imported and then hibernate resumes the system where the zpool is already imported and online. I guess it is probably the reason for my corrupted zpool.

Thanks @problame for your support.

**Comment by AttilaFueloep** [comment]
Well I can add to that. I'm hibernating regularly since a couple of years without any problem so far (knock on wood). I've root on zfs but boot and swap on luks on top of mdadm mirrors. I'm using mkinitcpio if that matters.

**Comment by luke-jr** [comment]
Maybe ZFS should refuse to mount read-write without the user forcing it if it believes it was mounted at hibernation? (I'm assuming a read-only mount won't change anything on disk...)

**Comment by bghira** [comment]
if you can help point out which docs might need updating to include these hints then we might make some progress on it

i know there's a few install guides that are hosted by the OpenZFS project that could be enhanced. each one could link to a page of caveats, which would mean just one spot to maintain them.

i would suggest that it be added to the man pages but ever since they were 'modernised' by breaking them out into separate pages, i have found them to be less useful and rarely search them for information that's now easier to find on Google.

**Comment by eblau** [comment]
@danielmorlock @problame wow, thanks for that hint on the initramfs scripts issue! I use Arch Linux and think that I'm in the same situation with the scripts importing the zpool and then decrypting swap on resume.

I store the swap partition's key in the zpool itself so on resume it does the following:

1. Runs "cryptsetup open" to prompt for the password to open the encrypted LUKS partition with the zpool on it.
2. Imports the zpool.
3. Invokes "cryptsetup open" to open the encrypted LUKS swap partition using the swap partition's key in the zpool.
4. Unmounts the root dataset and exports the zpool.
5. Resumes from hibernate.

I'm assuming that step 2 is the issue since the state of the zpool on disk could then differ the in-memory state in swap that we resume from. Would this work if the pool is imported read-only in step 2?

**Comment by problame** [comment]
@eblau yeah, that sound unhealthy.

> Maybe ZFS should refuse to mount read-write without the user forcing it if it believes it was mounted at hibernation? (I'm assuming a read-only mount won't change anything on disk...)

`zfs import` should  actually fail. I guess many scripts use `zpool import -f` to shoot themselves in the foot :)
@danielmorlock can you confirm it was using `zpool import -f`?

Regardless, I think -f shouldn't be sufficient to import a pool that was hibernated.
Idea:

Hibernation workflow:
* somehow get notified by the kernel that we're hibernating
  *  If I remember today's session correctly, `freeze_fs` and `thaw_fs` are not useful for this.
* generate a random *hibernation cookie*
* store the hibernation cookie in the in-DRAM spa_t, and somewhere on disk, let's say in the *MOS config*
* wait until the disk change txg has synced out
* but let all later txgs get stuck in transitioning from quiesce -> syncing 
  * implementation in txg.c, just prevent the transition from happening
* now allow hibernation. it must not be allowed before.

Resume workflow:
* Let initrd restore kernel threads and userland
* Somehow get notified from the kernel that we're resuming. That should be possible.
* Load the *MOS config* from disk
* Compare the hibernation cookie stored in the *MOS config* with the one we have in the (restored) DRAM
  * If they don't match, panic the kernel with the following message:
    ```
    zpool was used inbetween hibernation and resume
    ```
  * If they match, allow quiescing -> syncing transitions again.

To prevent accidental imports, we extend `zpool import` / `spa_import` such that they will fail by default a hibernation cookie is present in the MOS config.
This behavior can be overridden by *a new flag* `zpool import --discard-hibernation-state-and-fail-resume`.

Thoughts on this design, @behlendorf ?

**Comment by eblau** [comment]
> @eblau yeah, that sound unhealthy.
> 
> > Maybe ZFS should refuse to mount read-write without the user forcing it if it believes it was mounted at hibernation? (I'm assuming a read-only mount won't change anything on disk...)
> 
> `zfs import` should actually fail. I guess many scripts use `zpool import -f` to shoot themselves in the foot :) @danielmorlock can you confirm it was using `zpool import -f`?

My initcpio script is not using `-f`. Here are the exact commands it runs:
```
    modprobe zfs
    mkdir /crypto_key_device
    zpool import -d /dev/mapper -R /crypto_key_device zroot
    cryptsetup open --type=luks --key-file /crypto_key_device/etc/swapkeyfile --allow-discards /dev/nvme0n1p3 cryptswap
    zfs unmount -f /crypto_key_device
    zpool export zroot
```

**Comment by danielmorlock** [comment]
> zfs import should actually fail. I guess many scripts use zpool import -f to shoot themselves in the foot :)
> @danielmorlock can you confirm it was using zpool import -f?

Im pretty sure, that `-f` was not used: `man genkernel` says:
```
       dozfs[=cache,force]
           Scan for bootable ZFS pools on bootup. Optionally use cachefile or force import if necessary or perform both actions.
```

My kernel command line was:
```
options dozfs crypt_root=UUID=612a36bf-607c-4c8f-8dfd-498b87ea6b7f crypt_swap=UUID=8d173ef7-2af5-4ae5-9b7f-ad06985b1dd0 root=ZFS=rpool_ws1/system/root resume=UUID=74ef965e-688b-495d-95b4-afc449c15750 systemd.unified_cgroup_hierarchy=0
```
In the initrd phase, the zpool is tried to import before resuming from suspend to disk. So this is equal to import and already imported zpool from a different kernel. Is it?  Does ZFS track that it is already imported (from another system)?

**Comment by bghira** [comment]
> I'm assuming that step 2 is the issue since the state of the zpool on disk could then differ from the in-memory state in swap that we resume from. Would this work if the pool is imported read-only in step 2?

@behlendorf would be best to answer but from past discussions i recall him saying this would need to be very carefully handled (and might not be possible to, since many symbols are GPL-only surrounding the hibernation code, possibly why nvidia's implementation sucks as well) as it could lead to undefined behaviour and crashes, as the ZFS code is currently written.

**Comment by AttilaFueloep** [comment]
@problame 

That design sounds reasonable to me.

@eblau
> I'm assuming that step 2 is the issue

Yes, definitively. Once I wrecked a pool beyond repair by accidentally resuming. I wrote to it between hibernation and resume from a different system and after the resume things went south. But even an import/export cycle alone will change the pool so it won't match the state which is stored in the hibernation image. 

> Would this work if the pool is imported read-only in step 2?

IIRC there was an issue with read only import modifying some state of the pool. Not sure what the current situation is.

@danielmorlock 
> In the initrd phase, the zpool is tried to import before resuming from suspend to disk. So this is equal to import an already imported zpool from a different kernel. Is it

Yes.

> Does ZFS track that it is already imported (from another system)?

Not by default, MMP (Multi-modifier protection aka multihost) handles that but I can't tell if it would work in this case.

**Comment by eblau** [comment]
@Greek64 
I will check out the `decrypt_derived` and `decrypt_keyctl` suggestions, thank you. I do indeed do 2-stage LUKS unlocking to avoid typing a password twice.

I do it this way due to ignorance. :)  I researched LUKS on Arch Linux wiki pages and implemented that using the 2-stage unlock approach and then added hibernate/resume later without recognizing the bad interaction between the two.

Definitely it makes more sense to decrypt the swap before the root. That's why when I saw the explanation from @danielmorlock and @problame, I immediately realized the error of my ways.

Sorry for troubling the ZFS folks with this issue. The subject of this issue made me think that it was some ZFS support missing. Crazy thing is that I hibernated for like 2 years every day using this approach and only hit zpool corruption like 3 times. Luckily I take backups religiously and never lost much due to the magic of zfs send/receive.

**Comment by danielmorlock** [comment]
I've opened a bug ticket for genkernel: https://bugs.gentoo.org/827281

**Comment by problame** [comment]
We should close this issue since @behlendorf 's [original motivation](https://github.com/openzfs/zfs/issues/260#issue-980571) is misleading people into believing freeze & thaw are related to or required for supporting hibernation.

Barring some uncertainty from my side about in-flight IOs, my current understanding is that it's safe to hibernate a system with an imported pool if and only if the swap space into which the hibernation image is written resides outside of the zpool.
As [@danielmorlock and I figured out](https://github.com/openzfs/zfs/issues/260#issuecomment-978143047) it's a **very brittle** process though, since ZFS has no safeguards if your initrd scripts accidentally import the zpool on boot-to-resume.
I have outlined a design to improve this situation [above](https://github.com/openzfs/zfs/issues/260#issuecomment-978211802) and will create a new issue for it.

A few words on `freeze_fs` / `unfreeze_fs`, since they have been mentioned in this thread. The idea behind these callbacks is to  block out all VFS ops **to a single `struct super_block`** through an rwlock. Here's the [kernel function `freeze_super`](https://elixir.bootlin.com/linux/v5.10.82/source/fs/super.c#L1711) that is invoked from the ioctl, if `freeze_fs != NULL`.
As I understand it, the idea is that, with a one-super-block-per-bdev type of filesystem like XFS or Ext4, userspace can freeze the filesystem using the ioctl, then create a block-level snapshot or backup, then thaw it again using another ioctl.
If that understanding is correct and an exhaustive description of the use case, then I believe it is ill-advised to implement the callbacks for ZFS, since other datasets (= other super blocks) will continue to do IO to the same zpool. And even if userspace is careful to freeze all datasets, the Linux VFS isn't aware of the ZFS management operations that perform IO to the pool (send/recv, properties, ...).

Note that there are also the super block ops `freeze_super` / `thaw_super` (yep, confusing naming). A filesystem can implement these _instead of_ `freeze_fs` / `unfreeze_fs` if it wants to implement the infrastructure for locking out VFS ops itself instead of using the one provided by the [`freeze_super` function]((https://elixir.bootlin.com/linux/v5.10.82/source/fs/super.c#L1711)).

Note also: btrfs. It's the mainline filesystem most similar to ZFS with regard to pooled storage (multiple blockdevs!) and multiple super blocks on top. Btrfs implements `freeze_fs`/`unfreeze_fs`, but in `freeze_fs`. But the btrfs secret sauce is that a single btrfs _filesystem_ (= pool in ZFS terms) only has a single `struct suber_block` - the subvolumes (= ZPL datasets in ZFS terms) are implemented through `mount_subtree`.

**Comment by problame** [comment]
Maybe one more remark regarding **just how brittle hibernation currently is**:
hibernation does a system-wide [`sync` prior to "freezing" userspace and kernel threads](https://elixir.bootlin.com/linux/v5.10.82/source/kernel/power/hibernate.c#L730).
But there is no transactionality here, i.e., userspace and kernel threads can continue to dirty DMU state, issue and execute ZIOs to the pool, until they are "frozen" (sigh, naming, it should be called "descheduled for hibernation").

What does this mean?
* If we are able to successfully resume, everything should be ok, IF AND ONLY IF no zio's have been dropped on the floor during hibernation. Otherwise, those lost zios might later look a lot like data corruption. *Later* could be a lot later, if the hibernate image had the data cached in the ARC.
  * I haven't had time to look into this yet. Maybe the kernel quiesces the IO stack is quiesced before threads, I don't know. I wouldn't want to place a bet on it.
* If we are unable to resume, the situation is like a hard crash shortly after txg sync / ZIL write.
  * In theory, the pool should be safe to import. You might lose a few seconds worth of data. But since the kernel does the equivalent of a `sync` during hibernation, you probably didn't lose much in practice. Sync write guarantees should continue to hold as well. 
  * In practice, this still exercises a non-happy code path.

**Comment by behlendorf** [comment]
@problame your general design makes good sense to me.  It's been a while since I looked at this (years), but I agree it should largely be a matter of suspending the txg engine when hibernating and then resuming it after the system state has been restored.

> generate a random hibernation cookie
> store the hibernation cookie in the in-DRAM spa_t, and somewhere on disk, let's say in the MOS config

One avenue you may want to explore here is leveraging the existing `zio_suspend()` / `zio_resume()` machinery.  Fundamentally this code implements the logic to cleanly halt the txg engine, including suspending the zio pipeline and halting any inflight zios.  When resuming the pipeline is restarted and any halted zio's are resubmitted and allowed to proceed as if nothing happened.  Today this is what's used to suspended the pool when, due to drive failures, there's no longer sufficient redundancy to allow the pool to continue operation.  But I could image extending so the pool could be intentionally put in this suspended state for hibernation and a cookie stored.

Opening a new issue to track this work sounds like a good idea to me.  I'd just suggest that we somehow reference this old issue from the new one to make it easy to find.  I don't mind closing this one at all once we have a replacement to continue the discussion.

**Comment by problame** [comment]
@behlendorf I have created two follow-up issues:

- https://github.com/openzfs/zfs/issues/12842
- https://github.com/openzfs/zfs/issues/12843

I think you can close this issue now.

**Comment by alek-p** [comment]
possible related to https://github.com/openzfs/zfs/issues/13879

**Comment by nwg** [comment]
> Maybe one more remark regarding **just how brittle hibernation currently is**: hibernation does a system-wide [`sync` prior to "freezing" userspace and kernel threads](https://elixir.bootlin.com/linux/v5.10.82/source/kernel/power/hibernate.c#L730). But there is no transactionality here, i.e., userspace and kernel threads can continue to dirty DMU state, issue and execute ZIOs to the pool, until they are "frozen" (sigh, naming, it should be called "descheduled for hibernation").
> 

so what would a correct and performant fix look like then? Would it be ok to create a global lock on DMU operations and ZIO, preventing any ZIO or DMU from finishing, until the deschedule happens,
